mod-source-record-manager-server/src/main/java/org/folio/services/afterprocessing/InstanceProcessingServiceImpl.java

```diff
@@ -75,68 +63,109 @@ public InstanceProcessingServiceImpl(@Autowired JobExecutionSourceChunkDao jobEx
-    if (record.getParsedRecord() != null && record.getParsedRecord().getContent() != null) {
-      try {
+  private List<Pair<Record, Instance>> mapRecordsToInstances(List<Record> records) {
+    if (CollectionUtils.isEmpty(records)) {
+      return Collections.emptyList();
+    }
+    final RecordToInstanceMapper mapper = RecordToInstanceMapperBuilder.buildMapper(RecordFormat.getByDataType(getRecordsType(records)));
+    return records.parallelStream()
+      .map(record -> mapRecordToInstance(mapper, record))
+      .filter(Objects::nonNull)
+      .collect(Collectors.toList());
+  }
+
+  /**
+   * @param mapper
+   * @param record
+   * @return
+   */
+  private Pair<Record, Instance> mapRecordToInstance(RecordToInstanceMapper mapper, Record record) {
+    try {
+      if (record.getParsedRecord() != null && record.getParsedRecord().getContent() != null) {
         Instance instance = mapper.mapRecord(new JsonObject(record.getParsedRecord().getContent().toString()));
-        return postInstance(instance, params).compose(instanceId -> {
-          if (Record.RecordType.MARC.equals(record.getRecordType())) {
-            additionalInstanceFieldsUtil.addInstanceIdToMarcRecord(record, instanceId);
-          }
-          return Future.succeededFuture();
-        });
-      } catch (Exception exception) {
-        String errorMessage =
-          String.format("Can not process given Record for Instance. Cause: '%s'. Exception: '%s'", exception.getMessage(), exception);
-        LOGGER.error(errorMessage);
-        return Future.failedFuture(errorMessage);
+        return Pair.of(record, instance);
       }
+    } catch (Exception exception) {
+      String errorMessage = String.format("Can not map a given Record to Instance. Cause: '%s'. Exception: '%s'", exception.getMessage(), exception);
+      LOGGER.error(errorMessage);
     }
-    // IF given Record is ERROR record - return succeeded future
-    return Future.succeededFuture();
+    return null;
   }
 
   /**
-   * Creates Inventory Instance
+   * Sends given collection of Instances to mod-inventory
    *
-   * @param instance Instance entity serves as request body to POST request to Inventory
-   * @param params   params enough to connect ot OKAPI
-   * @return future with Instance id
+   * @param instances collection of Instances
+   * @param params    Okapi connection params
+   * @return future
    */
-  private Future<String> postInstance(Instance instance, OkapiConnectionParams params) {
-    Future<String> future = Future.future();
-    RestUtil.doRequest(params, INVENTORY_URL, HttpMethod.POST, instance).setHandler(responseResult -> {
+  private Future<Void> postInstances(List<Instance> instances, OkapiConnectionParams params) {
+    if (CollectionUtils.isEmpty(instances)) {
+      return Future.succeededFuture();
+    }
+    Future<Void> future = Future.future();
+    RestUtil.doRequest(params, INVENTORY_URL, HttpMethod.POST, instances).setHandler(responseResult -> {
```

Comment:
As I remember, batch post request body is instances object with totalCount field and list of instances